opm/simulators/linalg/FlexibleSolver.hpp

```diff
@@ -0,0 +1,175 @@
+class FlexibleSolver : public Dune::SolverWithUpdate<VectorTypeT, VectorTypeT>
+{
+public:
+    using MatrixType = MatrixTypeT;
+    using VectorType = VectorTypeT;
+
+    /// Create a sequential solver.
+    FlexibleSolver(const boost::property_tree::ptree& prm, const MatrixType& matrix)
+    {
+        init(prm, matrix, Dune::Amg::SequentialInformation());
+    }
+
+    /// Create a parallel solver (if Comm is e.g. OwnerOverlapCommunication).
+    template <class Comm>
+    FlexibleSolver(const boost::property_tree::ptree& prm, const MatrixType& matrix, const Comm& comm)
+    {
+        init(prm, matrix, comm);
+    }
+
+    virtual void apply(VectorType& x, VectorType& rhs, Dune::InverseOperatorResult& res) override
+    {
+        linsolver_->apply(x, rhs, res);
+    }
+
+    virtual void apply(VectorType& x, VectorType& rhs, double reduction, Dune::InverseOperatorResult& res) override
+    {
+        linsolver_->apply(x, rhs, reduction, res);
+    }
+
+    virtual void updatePreconditioner() override
+    {
+        preconditioner_->update();
+    }
+
+    virtual Dune::SolverCategory::Category category() const override
+    {
+        return linearoperator_->category();
+    }
+
+private:
+    using AbstractOperatorType = Dune::AssembledLinearOperator<MatrixType, VectorType, VectorType>;
+    using AbstractPrecondType = Dune::PreconditionerWithUpdate<VectorType, VectorType>;
+    using AbstractScalarProductType = Dune::ScalarProduct<VectorType>;
+    using AbstractSolverType = Dune::InverseOperator<VectorType, VectorType>;
+
+    // Machinery for making sequential or parallel operators/preconditioners/scalar products.
+    template <class Comm>
+    void initOpPrecSp(const MatrixType& matrix, const boost::property_tree::ptree& prm, const Comm& comm)
+    {
+        // Parallel case.
+        using ParOperatorType = Dune::OverlappingSchwarzOperator<MatrixType, VectorType, VectorType, Comm>;
+        auto linop = std::make_shared<ParOperatorType>(matrix, comm);
+        linearoperator_ = linop;
+        preconditioner_
+            = Dune::PreconditionerFactory<ParOperatorType, Comm>::create(*linop, prm.get_child("preconditioner"), comm);
+        scalarproduct_ = Dune::createScalarProduct<VectorType, Comm>(comm, linearoperator_->category());
+    }
+
+    void initOpPrecSp(const MatrixType& matrix, const boost::property_tree::ptree& prm, const Dune::Amg::SequentialInformation&)
+    {
+        // Sequential case.
+        using SeqOperatorType = Dune::MatrixAdapter<MatrixType, VectorType, VectorType>;
+        auto linop = std::make_shared<SeqOperatorType>(matrix);
+        linearoperator_ = linop;
+        preconditioner_ = Dune::PreconditionerFactory<SeqOperatorType>::create(*linop, prm.get_child("preconditioner"));
+        scalarproduct_ = std::make_shared<Dune::SeqScalarProduct<VectorType>>();
+    }
+
+    void initSolver(const boost::property_tree::ptree& prm)
+    {
+        const double tol = prm.get<double>("tol");
```

Comment:
Partly true. If the `--linear-solver-configuration-json-file` parameter is defaulted ("none") then a parameter tree is constructed that will give the default (block bicgstab + ilu0) solver and honor the usual command-line parameters (`--linear-solver-reduction`, `--linear-solver-maxiter` and `--linear-solver-verbosity`). See setupPropertyTree.cpp for the implementation.

If however you give a JSON file configuration, that will override anything else on the command line. In part because given a hierarchy of methods, each of them need their own tolerances (for example I run CPR with 1e-1 tolerance for the pressure solve). There is no obvious way to do this mapping in general, so I don't. I do not think the current parameter system is quite flexible enough to handle a general tree with a run-time dependent structure, which is why the boost::property_tree::ptree class has been used.